pkg/providers/instancetype/offering/provider.go

```diff
@@ -0,0 +1,274 @@
+				instanceTypeLabel: it.Name,
+				capacityTypeLabel: of.Requirements.Get(karpv1.CapacityTypeLabelKey).Any(),
+				zoneLabel:         of.Requirements.Get(corev1.LabelTopologyZone).Any(),
+			})
+			InstanceTypeOfferingPriceEstimate.Set(of.Price, map[string]string{
+				instanceTypeLabel: it.Name,
+				capacityTypeLabel: of.Requirements.Get(karpv1.CapacityTypeLabelKey).Any(),
+				zoneLabel:         of.Requirements.Get(corev1.LabelTopologyZone).Any(),
+			})
+		}
+
+		its = append(its, &cloudprovider.InstanceType{
+			Name:         it.Name,
+			Requirements: it.Requirements,
+			Offerings:    offerings,
+			Capacity:     it.Capacity,
+			Overhead:     it.Overhead,
+		})
+	}
+	return its
+}
+
+//nolint:gocyclo
+func (p *DefaultProvider) createOfferings(
+	it *cloudprovider.InstanceType,
+	nodeClass *v1.EC2NodeClass,
+	allZones sets.Set[string],
+	subnetZones map[string]string,
+	snapshotUUID *types.UID,
+) cloudprovider.Offerings {
+	itZones := sets.New(it.Requirements.Get(corev1.LabelTopologyZone).Values()...)
+
+	offerings := []cloudprovider.Offering{}
+	for zone := range allZones {
+		for _, capacityType := range it.Requirements.Get(karpv1.CapacityTypeLabelKey).Values() {
+			// Reserved capacity types are constructed separately, skip them for now.
+			if capacityType == karpv1.CapacityTypeReserved {
+				continue
+			}
+
+			isUnavailable := p.unavailableOfferings.IsUnavailable(it.Name, zone, capacityType)
+			_, hasSubnetZone := subnetZones[zone]
+			var price float64
+			var hasPrice bool
+			switch capacityType {
+			case karpv1.CapacityTypeOnDemand:
+				price, hasPrice = p.pricingProvider.OnDemandPrice(ec2types.InstanceType(it.Name))
+			case karpv1.CapacityTypeSpot:
+				price, hasPrice = p.pricingProvider.SpotPrice(ec2types.InstanceType(it.Name), zone)
+			default:
+				panic(fmt.Sprintf("invalid capacity type %q in requirements for instance type %q", capacityType, it.Name))
+			}
+			offering := cloudprovider.Offering{
+				Requirements: scheduling.NewRequirements(
+					scheduling.NewRequirement(karpv1.CapacityTypeLabelKey, corev1.NodeSelectorOpIn, capacityType),
+					scheduling.NewRequirement(corev1.LabelTopologyZone, corev1.NodeSelectorOpIn, zone),
+					scheduling.NewRequirement(v1.LabelCapacityReservationID, corev1.NodeSelectorOpDoesNotExist),
+				),
+				Price:     price,
+				Available: !isUnavailable && hasPrice && itZones.Has(zone) && hasSubnetZone,
+			}
+			if id, ok := subnetZones[zone]; ok {
+				offering.Requirements.Add(scheduling.NewRequirement(v1.LabelTopologyZoneID, corev1.NodeSelectorOpIn, id))
+			}
+			offerings = append(offerings, offering)
+		}
+	}
+
+	for i := range nodeClass.Status.CapacityReservations {
+		if nodeClass.Status.CapacityReservations[i].InstanceType != it.Name {
+			continue
+		}
+		reservation := &nodeClass.Status.CapacityReservations[i]
+
+		isUnavailable := p.unavailableOfferings.IsReservationUnavailable(reservation.ID)
```

Comment:
I know that this is marked as False all the time right now -- is this check even necessary since we should just have the AvailableInstanceCount of the reservation be 0 if it's not available

Reply:
The logic here has changed quite a lot and I'll note it down here - we're no longer discovering cr availability from the NodeClass status, it's stored in an in-memory cache in the capacity reservation provider. This cache is synced with the EC2 API whenever a NodeClass refreshes its status, and it also accounts for launches, terminations, and ICE errors. We now rely on the entirely instead of the unavailable cache.